File: apps/android/Components/NostrContent/NostrContentView.kt
// Requires: compose-ui, compose-foundation, compose-material3,
// androidx.compose.material:material-icons-extended (for ExpandMore,
// HelpOutline, PlayArrow, VolumeUp). Pulls `compose/content-media-grid` and
// `compose/content-quote-card` as registry deps. Kotlin 1.9+.
//
// Compose mirror of the SwiftUI `NostrContentView`. Walks a
// `ContentTreeWire`, flattens the arena into block-level groups via
// `nostrContentGroups`, and renders each block (paragraph / heading / media /
// code / list / quote / rule / image / event-ref / placeholder).
//
// Data injection contract:
//   - Theming + tap callbacks come from `LocalNostrContentRenderer`
//     (see `compose/content-core`).
//   - Mention display labels are provided by the app via `mentionLabel`.
//   - Quote / embed cards are app-owned: pass a `quoteCardProvider` closure
//     that returns a `NostrQuoteCardModel` for an event-ref URI. When `null`,
//     the view falls back to a `.Collapsed` quote card.
//
// Inline runs are flattened into a single `AnnotatedString` (with per-run
// styling) and shown as a `ClickableText` so tap-offset → annotation routing
// can dispatch the matching callback. Block nodes use `Column`.

package nmp.content

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.ClickableText
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material.icons.filled.HelpOutline
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.VolumeUp
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

@Composable
public fun NostrContentView(
    tree: ContentTreeWire,
    modifier: Modifier = Modifier,
    textStyle: TextStyle = MaterialTheme.typography.bodyLarge,
    mentionLabel: (WireNostrUri) -> String = ::defaultMentionLabel,
    quoteCardProvider: ((WireNostrUri) -> NostrQuoteCardModel?)? = null,
) {
    val groups = nostrContentGroups(tree)
    if (groups.isEmpty()) return

    Column(
        verticalArrangement = Arrangement.spacedBy(8.dp),
        modifier = modifier,
    ) {
        for (group in groups) {
            RenderGroup(
                group = group,
                tree = tree,
                textStyle = textStyle,
                mentionLabel = mentionLabel,
                quoteCardProvider = quoteCardProvider,
            )
        }
    }
}

public fun defaultMentionLabel(uri: WireNostrUri): String {
    val value = uri.primaryId
    if (value.length <= 12) return value
    return "${value.take(8)}…${value.takeLast(4)}"
}

// ---------------------------------------------------------------------------
// Block dispatch
// ---------------------------------------------------------------------------

@Composable
private fun RenderGroup(
    group: NostrContentGroup,
    tree: ContentTreeWire,
    textStyle: TextStyle,
    mentionLabel: (WireNostrUri) -> String,
    quoteCardProvider: ((WireNostrUri) -> NostrQuoteCardModel?)?,
) {
    when (group) {
        is NostrContentGroup.Inline -> InlineGroup(
            level = group.level,
            children = group.children,
            tree = tree,
            textStyle = textStyle,
            mentionLabel = mentionLabel,
        )
        is NostrContentGroup.MediaGroup -> MediaGroupBlock(
            urls = group.urls,
            kind = group.kind,
        )
        is NostrContentGroup.EventRefGroup -> EventRefBlock(
            uri = group.uri,
            quoteCardProvider = quoteCardProvider,
        )
        is NostrContentGroup.CodeBlockGroup -> CodeBlockBlock(
            info = group.info,
            body = group.body,
        )
        is NostrContentGroup.BlockQuoteGroup -> BlockQuoteBlock(
            children = group.children,
            tree = tree,
            textStyle = textStyle,
            mentionLabel = mentionLabel,
        )
        is NostrContentGroup.ListGroup -> ListBlock(
            orderedStart = group.orderedStart,
            items = group.items,
            tree = tree,
            textStyle = textStyle,
            mentionLabel = mentionLabel,
        )
        NostrContentGroup.RuleGroup -> RuleBlock()
        is NostrContentGroup.ImageGroup -> ImageBlock(
            alt = group.alt,
            title = group.title,
            src = group.src,
        )
        is NostrContentGroup.PlaceholderGroup -> PlaceholderChip(reason = group.reason)
    }
}

// ---------------------------------------------------------------------------
// Inline rendering (Text + Text concatenation → AnnotatedString)
// ---------------------------------------------------------------------------

@Composable
private fun InlineGroup(
    level: NostrContentInlineLevel,
    children: List<UInt>,
    tree: ContentTreeWire,
    textStyle: TextStyle,
    mentionLabel: (WireNostrUri) -> String,
) {
    val renderer = LocalNostrContentRenderer.current
    val annotated = buildAnnotatedString {
        for (child in children) {
            appendInline(
                index = child,
                tree = tree,
                renderer = renderer,
                mentionLabel = mentionLabel,
            )
        }
    }
    val effectiveStyle = when (level) {
        NostrContentInlineLevel.Paragraph -> textStyle.copy(color = renderer.textColor)
        is NostrContentInlineLevel.Heading -> headingStyle(level.level).copy(color = renderer.textColor)
    }
    ClickableText(
        text = annotated,
        style = effectiveStyle,
        modifier = Modifier.fillMaxWidth(),
        onClick = { offset ->
            dispatchInlineTap(annotated, offset, renderer)
        },
    )
}

/**
 * Append one arena node's inline projection to the [AnnotatedString] builder.
 * Recursive children (emphasis / strong / link / heading / paragraph) are
 * walked here so the whole inline subtree collapses into the running text.
 * Block-level nodes (`list`, `code_block`, `rule`, `media`, `placeholder`)
 * appearing inside an inline group emit nothing rather than break flow.
 */
private fun AnnotatedString.Builder.appendInline(
    index: UInt,
    tree: ContentTreeWire,
    renderer: NostrContentRenderer,
    mentionLabel: (WireNostrUri) -> String,
) {
    if (index == NOSTR_CONTENT_NEWLINE_SENTINEL) {
        append('\n')
        return
    }
    val node = tree.nodeAt(index) ?: return
    when (node) {
        is WireNode.Text -> append(node.text)
        is WireNode.Mention -> {
            val label = mentionLabel(node.uri)
            withAnnotationScope(MENTION_ANNOTATION, node.uri.primaryId) {
                withStyleScope(
                    SpanStyle(
                        color = renderer.mentionColor,
                        fontWeight = FontWeight.Bold,
                    ),
                ) { append("@$label") }
            }
        }
        is WireNode.EventRef -> {
            val short = shortEntity(node.uri.primaryId)
            withAnnotationScope(EVENT_REF_ANNOTATION, node.uri.primaryId) {
                withStyleScope(
                    SpanStyle(
                        color = renderer.linkColor,
                        fontWeight = FontWeight.Bold,
                    ),
                ) { append("↩ $short") }
            }
        }
        is WireNode.Hashtag -> {
            withAnnotationScope(HASHTAG_ANNOTATION, node.tag) {
                withStyleScope(
                    SpanStyle(
                        color = renderer.hashtagColor,
                        fontWeight = FontWeight.Bold,
                    ),
                ) { append("#${node.tag}") }
            }
        }
        is WireNode.Url -> {
            withAnnotationScope(LINK_ANNOTATION, node.url) {
                withStyleScope(SpanStyle(color = renderer.linkColor)) {
                    append(node.url)
                }
            }
        }
        is WireNode.Emoji -> append(":${node.shortcode}:")
        is WireNode.Invoice -> {
            withStyleScope(SpanStyle(color = renderer.linkColor)) {
                append("⚡ invoice")
            }
        }
        is WireNode.Emphasis -> withStyleScope(SpanStyle(fontStyle = FontStyle.Italic)) {
            for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        }
        is WireNode.Strong -> withStyleScope(SpanStyle(fontWeight = FontWeight.Bold)) {
            for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        }
        is WireNode.InlineCode -> withStyleScope(SpanStyle(fontFamily = FontFamily.Monospace)) {
            append(node.code)
        }
        is WireNode.Link -> {
            val href = node.href
            if (!href.isNullOrEmpty()) {
                withAnnotationScope(LINK_ANNOTATION, href) {
                    withStyleScope(
                        SpanStyle(
                            color = renderer.linkColor,
                            textDecoration = TextDecoration.Underline,
                        ),
                    ) {
                        for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
                    }
                }
            } else {
                for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
            }
        }
        is WireNode.Image -> {
            val alt = node.alt
            withStyleScope(SpanStyle(color = renderer.placeholderColor)) {
                append(if (alt.isEmpty()) "[image]" else "[$alt]")
            }
        }
        WireNode.SoftBreak -> append(' ')
        WireNode.HardBreak -> append('\n')
        is WireNode.Paragraph -> for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        is WireNode.Heading -> for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        is WireNode.BlockQuote -> for (child in node.children) appendInline(child, tree, renderer, mentionLabel)
        is WireNode.ListNode,
        is WireNode.CodeBlock,
        WireNode.Rule,
        is WireNode.Media,
        is WireNode.Placeholder -> { /* block-level — never inside an inline reduce */ }
    }
}

/** Resolve a tap offset against the annotations attached during inline build. */
private fun dispatchInlineTap(
    annotated: AnnotatedString,
    offset: Int,
    renderer: NostrContentRenderer,
) {
    annotated.getStringAnnotations(MENTION_ANNOTATION, offset, offset).firstOrNull()?.let {
        renderer.callbacks.onMentionTap(it.item)
        return
    }
    annotated.getStringAnnotations(EVENT_REF_ANNOTATION, offset, offset).firstOrNull()?.let {
        renderer.callbacks.onEventRefTap(it.item)
        return
    }
    annotated.getStringAnnotations(HASHTAG_ANNOTATION, offset, offset).firstOrNull()?.let {
        renderer.callbacks.onHashtagTap(it.item)
        return
    }
    annotated.getStringAnnotations(LINK_ANNOTATION, offset, offset).firstOrNull()?.let {
        renderer.callbacks.onLinkTap(it.item)
    }
}

// Annotation tags used to round-trip per-run identifiers from
// AnnotatedString into the tap-offset dispatcher. Centralized so the
// builder and the dispatcher cannot drift.
private const val MENTION_ANNOTATION = "nmp:mention"
private const val EVENT_REF_ANNOTATION = "nmp:event_ref"
private const val HASHTAG_ANNOTATION = "nmp:hashtag"
private const val LINK_ANNOTATION = "nmp:link"

/** Push a [SpanStyle] for the duration of [block] then pop it. Avoids the
 *  `androidx.compose.ui.text.withStyle` import to keep this file's API
 *  surface explicit. */
private inline fun AnnotatedString.Builder.withStyleScope(
    style: SpanStyle,
    block: AnnotatedString.Builder.() -> Unit,
) {
    val index = pushStyle(style)
    try {
        block()
    } finally {
        pop(index)
    }
}

/** Push a string annotation for the duration of [block] then pop it. */
private inline fun AnnotatedString.Builder.withAnnotationScope(
    tag: String,
    annotation: String,
    block: AnnotatedString.Builder.() -> Unit,
) {
    val index = pushStringAnnotation(tag, annotation)
    try {
        block()
    } finally {
        pop(index)
    }
}

private fun headingStyle(level: UByte): TextStyle {
    return when (level.toInt()) {
        1 -> TextStyle(fontSize = 30.sp, fontWeight = FontWeight.Bold)
        2 -> TextStyle(fontSize = 26.sp, fontWeight = FontWeight.Bold)
        3 -> TextStyle(fontSize = 22.sp, fontWeight = FontWeight.SemiBold)
        4 -> TextStyle(fontSize = 19.sp, fontWeight = FontWeight.SemiBold)
        5 -> TextStyle(fontSize = 17.sp, fontWeight = FontWeight.SemiBold)
        else -> TextStyle(fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
    }
}

private fun shortEntity(value: String): String {
    if (value.length <= 12) return value
    return "${value.take(8)}…${value.takeLast(4)}"
}

// ---------------------------------------------------------------------------
// Block builders
// ---------------------------------------------------------------------------

@Composable
private fun MediaGroupBlock(urls: List<String>, kind: MediaKind) {
    when (kind) {
        MediaKind.Image -> {
            val nonEmpty = urls.filter { it.isNotEmpty() }
            if (nonEmpty.isNotEmpty()) {
                NostrMediaGrid(imageUrls = nonEmpty)
            }
        }
        MediaKind.Video, MediaKind.Audio -> {
            val first = urls.firstOrNull()
            if (first != null) {
                MediaRow(url = first, isAudio = kind == MediaKind.Audio)
            }
        }
    }
}

@Composable
private fun MediaRow(url: String, isAudio: Boolean) {
    val renderer = LocalNostrContentRenderer.current
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(renderer.codeBackgroundColor)
            .clickable { renderer.callbacks.onLinkTap(url) }
            .padding(12.dp),
    ) {
        Icon(
            imageVector = if (isAudio) Icons.Filled.VolumeUp else Icons.Filled.PlayArrow,
            contentDescription = null,
            tint = renderer.linkColor,
        )
        Text(
            text = url.substringAfterLast('/').ifEmpty { url },
            color = renderer.secondaryTextColor,
            fontFamily = FontFamily.Monospace,
            maxLines = 1,
            style = MaterialTheme.typography.bodySmall,
        )
    }
}

@Composable
private fun EventRefBlock(
    uri: WireNostrUri,
    quoteCardProvider: ((WireNostrUri) -> NostrQuoteCardModel?)?,
) {
    // Variant selection mirrors SwiftUI: provider missing → .Collapsed,
    // provider hit → .Rich, provider miss → .Missing.
    val providedModel = quoteCardProvider?.invoke(uri)
    val variant: NostrQuoteCardVariant
    val model: NostrQuoteCardModel
    when {
        quoteCardProvider == null -> {
            variant = NostrQuoteCardVariant.Collapsed
            model = NostrQuoteCardModel(id = uri.primaryId, unresolvedUri = uri.uri)
        }
        providedModel != null -> {
            variant = NostrQuoteCardVariant.Rich
            model = providedModel
        }
        else -> {
            variant = NostrQuoteCardVariant.Missing
            model = NostrQuoteCardModel(id = uri.primaryId, unresolvedUri = uri.uri)
        }
    }
    val renderer = LocalNostrContentRenderer.current
    NostrQuoteCard(
        model = model,
        variant = variant,
        onTap = { renderer.callbacks.onEventRefTap(uri.primaryId) },
    )
}

@Composable
private fun CodeBlockBlock(info: String?, body: String) {
    val renderer = LocalNostrContentRenderer.current
    Column(
        verticalArrangement = Arrangement.spacedBy(4.dp),
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(6.dp))
            .background(renderer.codeBackgroundColor)
            .padding(10.dp),
    ) {
        if (!info.isNullOrEmpty()) {
            Text(
                text = info,
                color = renderer.secondaryTextColor,
                fontFamily = FontFamily.Monospace,
                style = MaterialTheme.typography.labelSmall,
            )
        }
        Text(
            text = body,
            color = renderer.textColor,
            fontFamily = FontFamily.Monospace,
            modifier = Modifier.fillMaxWidth(),
        )
    }
}

@Composable
private fun BlockQuoteBlock(
    children: List<UInt>,
    tree: ContentTreeWire,
    textStyle: TextStyle,
    mentionLabel: (WireNostrUri) -> String,
) {
    val renderer = LocalNostrContentRenderer.current
    val annotated = buildAnnotatedString {
        for (child in children) {
            appendInline(child, tree, renderer, mentionLabel)
        }
    }
    Row(
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        modifier = Modifier
            .fillMaxWidth()
            .height(IntrinsicSize.Min)
            .padding(vertical = 4.dp),
    ) {
        Box(
            modifier = Modifier
                .width(3.dp)
                .fillMaxHeight()
                .background(renderer.quoteBorderColor),
        )
        Text(
            text = annotated,
            color = renderer.secondaryTextColor,
            style = textStyle.copy(fontStyle = FontStyle.Italic),
            modifier = Modifier.fillMaxWidth(),
        )
    }
}

@Composable
private fun ListBlock(
    orderedStart: ULong?,
    items: List<List<UInt>>,
    tree: ContentTreeWire,
    textStyle: TextStyle,
    mentionLabel: (WireNostrUri) -> String,
) {
    val renderer = LocalNostrContentRenderer.current
    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
        items.forEachIndexed { offset, children ->
            val annotated = buildAnnotatedString {
                for (child in children) {
                    appendInline(child, tree, renderer, mentionLabel)
                }
            }
            Row(
                verticalAlignment = Alignment.Top,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                modifier = Modifier.fillMaxWidth(),
            ) {
                Text(
                    text = marker(orderedStart, offset),
                    color = renderer.secondaryTextColor,
                    style = textStyle,
                )
                Text(
                    text = annotated,
                    color = renderer.textColor,
                    style = textStyle,
                    modifier = Modifier.fillMaxWidth(),
                )
            }
        }
    }
}

private fun marker(orderedStart: ULong?, offset: Int): String {
    if (orderedStart != null) {
        return "${orderedStart + offset.toULong()}."
    }
    return "•"
}

@Composable
private fun RuleBlock() {
    val renderer = LocalNostrContentRenderer.current
    Spacer(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 4.dp)
            .height(1.dp)
            .background(renderer.quoteBorderColor),
    )
}

@Composable
private fun ImageBlock(alt: String, @Suppress("UNUSED_PARAMETER") title: String?, src: String?) {
    val renderer = LocalNostrContentRenderer.current
    if (!src.isNullOrEmpty()) {
        // `title` is reserved for a future accessibility label parity with the
        // SwiftUI implementation; the media grid currently only takes URLs.
        NostrMediaGrid(imageUrls = listOf(src))
        return
    }
    Text(
        text = if (alt.isEmpty()) "[image]" else "[$alt]",
        color = renderer.placeholderColor,
        style = MaterialTheme.typography.bodySmall,
    )
}

@Composable
private fun PlaceholderChip(reason: PlaceholderReason) {
    val renderer = LocalNostrContentRenderer.current
    val (label, icon) = when (reason) {
        PlaceholderReason.DepthLimit -> "Nested content collapsed" to Icons.Filled.ExpandMore
        PlaceholderReason.UnresolvedUri -> "Unresolved reference" to Icons.Filled.HelpOutline
    }
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        modifier = Modifier
            .clip(RoundedCornerShape(percent = 50))
            .background(renderer.codeBackgroundColor)
            .padding(horizontal = 8.dp, vertical = 4.dp),
    ) {
        Icon(imageVector = icon, contentDescription = null, tint = renderer.placeholderColor)
        Text(
            text = label,
            color = renderer.placeholderColor,
            style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.SemiBold),
        )
    }
}

File: apps/ios/Components/NostrContent/NostrContentView.swift
import AVKit
import SwiftUI
import UIKit

/// SwiftUI renderer for a `ContentTreeWire`. Walks `tree.roots`, flattens the
/// arena into block-level groups via `nostrContentGroups`, and renders each
/// block (paragraph/heading/media/code/list/quote/rule/image/event-ref/
/// placeholder) with SwiftUI `Text` concatenation for inline runs.
///
/// Data injection contract:
///   • Theming + tap callbacks come from `NostrContentRenderer` in the
///     SwiftUI environment (see `swiftui/content-core`).
///   • Mention display labels are provided by the app via `mentionLabel`.
///   • Quote/embed cards are app-owned: the app passes a `quoteCardProvider`
///     closure that returns a `NostrQuoteCardModel` for an event-ref URI.
///     When `nil`, the view falls back to a `.collapsed` quote card.
public struct NostrContentView: View {
    public var tree: ContentTreeWire
    public var font: Font
    public var mentionLabel: (NostrWireUri) -> String
    public var quoteCardProvider: ((NostrWireUri) -> NostrQuoteCardModel?)?

    @Environment(\.nostrContentRenderer) private var renderer

    public init(
        tree: ContentTreeWire,
        font: Font = .body,
        mentionLabel: @escaping (NostrWireUri) -> String = NostrContentView.defaultMentionLabel,
        quoteCardProvider: ((NostrWireUri) -> NostrQuoteCardModel?)? = nil
    ) {
        self.tree = tree
        self.font = font
        self.mentionLabel = mentionLabel
        self.quoteCardProvider = quoteCardProvider
    }

    public var body: some View {
        let groups = nostrContentGroups(tree)
        if groups.isEmpty {
            EmptyView()
        } else {
            VStack(alignment: .leading, spacing: 8) {
                ForEach(Array(groups.enumerated()), id: \.offset) { _, group in
                    groupView(group)
                }
            }
        }
    }

    // MARK: - Group dispatch

    @ViewBuilder
    private func groupView(_ group: NostrContentGroup) -> some View {
        switch group {
        case .inline(let level, let children):
            inlineGroup(level: level, children: children)
        case .media(let urls, let kind):
            mediaGroup(urls: urls, kind: kind)
        case .eventRef(let uri):
            eventRefView(uri)
        case .codeBlock(let info, let body):
            codeBlockView(info: info, body: body)
        case .blockQuote(let children):
            blockQuoteView(children: children)
        case .list(let orderedStart, let items):
            listView(orderedStart: orderedStart, items: items)
        case .rule:
            ruleView
        case .image(let alt, let title, let src):
            imageBlockView(alt: alt, title: title, src: src)
        case .placeholder(let reason):
            placeholderChip(reason: reason)
        }
    }

    @ViewBuilder
    private func inlineGroup(level: NostrContentInlineLevel, children: [UInt32]) -> some View {
        let concatenated = children.reduce(Text("")) { acc, child in
            acc + inlineText(child)
        }
        switch level {
        case .paragraph:
            concatenated
                .font(font)
                .foregroundStyle(renderer.textColor)
                .fixedSize(horizontal: false, vertical: true)
        case .heading(let lvl):
            concatenated
                .font(headingFont(for: lvl))
                .foregroundStyle(renderer.textColor)
                .fixedSize(horizontal: false, vertical: true)
        }
    }

    // MARK: - Inline text concatenation

    /// Convert one arena node into a `Text` value that can be `+`-concatenated
    /// with neighbouring runs. Recursive children (emphasis/strong/link/heading
    /// /paragraph) are walked here so the whole inline subtree collapses to a
    /// single `Text`.
    public func inlineText(_ index: UInt32) -> Text {
        if index == nostrContentNewlineSentinel { return Text("\n") }
        guard let node = tree.node(at: index) else { return Text("") }
        switch node {
        case .text(let value):
            return Text(value)
        case .mention(let uri):
            return Text("@\(mentionLabel(uri))")
                .foregroundStyle(renderer.mentionColor)
                .bold()
        case .eventRef(let uri):
            return Text("↩ \(shortEntity(uri.primaryId))")
                .foregroundStyle(renderer.linkColor)
                .bold()
        case .hashtag(let tag):
            return Text("#\(tag)")
                .foregroundStyle(renderer.hashtagColor)
                .bold()
        case .url(let value):
            return Text(value).foregroundStyle(renderer.linkColor)
        case .emoji(let shortcode, _):
            // Apps fill `renderer.emojiImages` from kind:0 / NIP-30 tag data
            // and the inline run renders the image directly. Empty dict
            // (the default) falls back to the literal `:shortcode:` text so
            // unwired apps still get a readable surface.
            if let img = renderer.emojiImages[shortcode] {
                return Text(Image(uiImage: img))
            }
            return Text(":\(shortcode):")
        case .invoice:
            return Text("⚡ invoice").foregroundStyle(renderer.linkColor)
        case .emphasis(let children):
            return children.reduce(Text("")) { $0 + inlineText($1).italic() }
        case .strong(let children):
            return children.reduce(Text("")) { $0 + inlineText($1).bold() }
        case .inlineCode(let value):
            return Text(value).font(.system(.body, design: .monospaced))
        case .link(let children, let href):
            let label = children.reduce(Text("")) { $0 + inlineText($1) }
            if let href, !href.isEmpty {
                return label
                    .foregroundStyle(renderer.linkColor)
                    .underline()
            }
            return label
        case .image(let alt, _, _):
            return Text(alt.isEmpty ? "[image]" : "[\(alt)]")
                .foregroundStyle(renderer.placeholderColor)
        case .softBreak:
            return Text(" ")
        case .hardBreak:
            return Text("\n")
        case .paragraph(let children),
             .heading(_, let children),
             .blockQuote(let children):
            return children.reduce(Text("")) { $0 + inlineText($1) }
        case .list, .codeBlock, .rule, .media, .placeholder:
            // Block-level — should never appear inside an inline reduce. Emit
            // nothing rather than break the text concatenation.
            return Text("")
        }
    }

    // MARK: - Block builders

    @ViewBuilder
    private func mediaGroup(urls: [String], kind: NostrMediaKind) -> some View {
        switch kind {
        case .image:
            let parsed = urls.compactMap(URL.init(string:))
            if !parsed.isEmpty {
                NostrMediaGrid(imageUrls: parsed)
            }
        case .video:
            // Inline playback via `AVKit.VideoPlayer` so video media nodes
            // render with native scrub / fullscreen controls. Audio stays on
            // the compact link-style row (no waveform UI in v1).
            if let first = urls.first.flatMap(URL.init(string:)) {
                VideoPlayer(player: AVPlayer(url: first))
                    .aspectRatio(16.0 / 9.0, contentMode: .fit)
                    .clipShape(RoundedRectangle(cornerRadius: 10))
            }
        case .audio:
            if let first = urls.first.flatMap(URL.init(string:)) {
                mediaRow(first, systemImage: "speaker.wave.2.fill")
            }
        }
    }

    private func mediaRow(_ url: URL, systemImage: String) -> some View {
        Button {
            renderer.callbacks.onLinkTap(url)
        } label: {
            HStack(spacing: 10) {
                Image(systemName: systemImage)
                    .font(.title2)
                    .foregroundStyle(renderer.linkColor)
                Text(url.lastPathComponent)
                    .font(.caption.monospaced())
                    .foregroundStyle(renderer.secondaryTextColor)
                    .lineLimit(1)
                Spacer(minLength: 0)
            }
            .padding(12)
            .frame(maxWidth: .infinity)
            .background(renderer.codeBackgroundColor, in: RoundedRectangle(cornerRadius: 8))
        }
        .buttonStyle(.plain)
    }

    private func eventRefView(_ uri: NostrWireUri) -> some View {
        // Variant selection is a three-branch decision so the `.missing`
        // variant is reachable:
        //   • no provider wired       → `.collapsed` (View quote affordance)
        //   • provider returned model → `.rich`
        //   • provider returned nil   → `.missing` (Content unavailable + URI)
        let providedModel = quoteCardProvider?(uri)
        let variant: NostrQuoteCardVariant
        let model: NostrQuoteCardModel
        if quoteCardProvider == nil {
            variant = .collapsed
            model = NostrQuoteCardModel(id: uri.primaryId, unresolvedUri: uri.uri)
        } else if let providedModel {
            variant = .rich
            model = providedModel
        } else {
            variant = .missing
            model = NostrQuoteCardModel(id: uri.primaryId, unresolvedUri: uri.uri)
        }
        return NostrQuoteCard(
            model: model,
            variant: variant,
            onTap: { renderer.callbacks.onEventRefTap(uri.primaryId) }
        )
    }

    private func codeBlockView(info: String?, body: String) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            if let info, !info.isEmpty {
                Text(info)
                    .font(.caption2.monospaced())
                    .foregroundStyle(renderer.secondaryTextColor)
            }
            Text(body)
                .font(.system(.callout, design: .monospaced))
                .foregroundStyle(renderer.textColor)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(10)
        .background(renderer.codeBackgroundColor, in: RoundedRectangle(cornerRadius: 6))
    }

    private func blockQuoteView(children: [UInt32]) -> some View {
        let text = children.reduce(Text("")) { $0 + inlineText($1) }
        return HStack(spacing: 10) {
            Rectangle()
                .fill(renderer.quoteBorderColor)
                .frame(width: 3)
            text
                .font(font.italic())
                .foregroundStyle(renderer.secondaryTextColor)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(.vertical, 4)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    @ViewBuilder
    private func listView(orderedStart: UInt64?, items: [[UInt32]]) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            ForEach(Array(items.enumerated()), id: \.offset) { offset, children in
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    Text(marker(orderedStart: orderedStart, offset: offset))
                        .font(font)
                        .foregroundStyle(renderer.secondaryTextColor)
                    children.reduce(Text("")) { $0 + inlineText($1) }
                        .font(font)
                        .foregroundStyle(renderer.textColor)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
        }
    }

    private func marker(orderedStart: UInt64?, offset: Int) -> String {
        if let orderedStart {
            return "\(orderedStart + UInt64(offset))."
        }
        return "•"
    }

    private var ruleView: some View {
        Rectangle()
            .fill(renderer.quoteBorderColor)
            .frame(height: 1)
            .padding(.vertical, 4)
    }

    private func imageBlockView(alt: String, title: String?, src: String?) -> some View {
        if let src, let url = URL(string: src) {
            return AnyView(
                NostrMediaGrid(imageUrls: [url])
                    .accessibilityLabel(title ?? alt)
            )
        }
        return AnyView(
            Text(alt.isEmpty ? "[image]" : "[\(alt)]")
                .font(.caption)
                .foregroundStyle(renderer.placeholderColor)
        )
    }

    private func placeholderChip(reason: NostrWirePlaceholderReason) -> some View {
        let label: String
        let icon: String
        switch reason {
        case .depthLimit:
            label = "Nested content collapsed"
            icon = "chevron.down.square"
        case .unresolvedUri:
            label = "Unresolved reference"
            icon = "questionmark.square.dashed"
        }
        return HStack(spacing: 6) {
            Image(systemName: icon)
                .foregroundStyle(renderer.placeholderColor)
            Text(label)
                .font(.caption.weight(.semibold))
                .foregroundStyle(renderer.placeholderColor)
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 4)
        .background(renderer.codeBackgroundColor, in: Capsule())
    }

    // MARK: - Defaults / helpers

    public nonisolated static func defaultMentionLabel(_ uri: NostrWireUri) -> String {
        let value = uri.primaryId
        guard value.count > 12 else { return value }
        return "\(value.prefix(8))…\(value.suffix(4))"
    }

    private func headingFont(for level: UInt8) -> Font {
        switch level {
        case 1: return .largeTitle.weight(.bold)
        case 2: return .title.weight(.bold)
        case 3: return .title2.weight(.semibold)
        case 4: return .title3.weight(.semibold)
        case 5: return .headline
        default: return .subheadline.weight(.semibold)
        }
    }

    private func shortEntity(_ value: String) -> String {
        guard value.count > 12 else { return value }
        return "\(value.prefix(8))…\(value.suffix(4))"
    }
}
